src/router.js

```diff
@@ -1,81 +1,76 @@
-export default function (render, options) {
-    var routes = options.view
-
-    function setLocation(data) {
-        render(match(routes, data))
-        history.pushState({}, "", data)
-    }
-
-    window.addEventListener("popstate", function () {
-        render(match(routes, location.pathname))
-    })
-
-    window.addEventListener("click", function (e) {
-        if (e.metaKey || e.shiftKey || e.ctrlKey || e.altKey) {
-            return
-        }
-
-        var target = e.target
-
-        while (target && target.localName !== "a") {
-            target = target.parentNode
-        }
-
-        if (target && target.host === location.host
-            && !target.hasAttribute("data-no-routing")) {
-
-            var element = document.querySelector(target.hash === "" ? element : target.hash)
-
-            if (element) {
-                element.scrollIntoView(true)
-            } else {
-                e.preventDefault()
-                setLocation(target.pathname)
-            }
-        }
-    })
-
-    render(match(routes, location.pathname))
-
-    function match(routes, path) {
-        for (var route in routes) {
-            var re = regexify(route), params = {}, match
+export default function (options) {
+	var routes = options.view
+
+	return {
+		model: {
+			router: {
+				location: location.pathname
+			}
+		},
+		effects: {
+			router: {
+				go: function (model, actions, data) {
+					actions.router.setLocation(data)
+					history.pushState({}, "", data)
+				}
+			}
+		},
+		reducers: {
+			router: {
+				setLocation: function (model, data) {
+					model.router.location = data
+					return model
+				}
+			}
+		},
+		subscriptions: [
+			function (_, actions) {
+				window.addEventListener("popstate", function () {
+					actions.router.setLocation(location.pathname)
+				})
+			}
+		],
+		hooks: {
+			onRender: function (model, view) {
+				return match(routes, model.router.location)
+			}
+		}
+	}
+}
 
-            if (route === "*") {
-                continue
-            }
+function match(routes, path) {
+	for (var route in routes) {
+		if (route === "*") {
+			continue
+		}
 
-            path.replace(new RegExp(re.re, "g"), function () {
-                for (var i = 1; i < arguments.length - 2; i++) {
-                    params[re.keys.shift()] = arguments[i]
-                }
+		var re = regexify(route), params = {}, match
 
-                match = function (model, actions) {
-                    actions.setLocation = setLocation
-                    return routes[route](model, actions, params)
-                }
-            })
+		path.replace(new RegExp(re.re, "g"), function () {
+			for (var i = 1; i < arguments.length - 2; i++) {
+				params[re.keys.shift()] = arguments[i]
+			}
 
-            if (match) {
-                return match
-            }
-        }
+			match = function (model, actions) {
+				return routes[route](model, actions, params)
```

Comment:
This should actually also improve performance since `match` doesn't need to be executed on **every** render and instead just on every `go`. I'd not be surprised if this gives us `-0.1` or more relative performance.